oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/RegisteredClient.java

```diff
@@ -498,8 +490,9 @@ private RegisteredClient create() {
 			registeredClient.clientSecret = this.clientSecret;
 			registeredClient.clientSecretExpiresAt = this.clientSecretExpiresAt;
 			registeredClient.clientName = this.clientName;
-			registeredClient.clientAuthenticationMethods = Collections.unmodifiableSet(
-					new HashSet<>(this.clientAuthenticationMethods));
+			registeredClient.clientAuthenticationMethods = this.clientAuthenticationMethods.stream()
+					.map(this::upgradeDeprecatedAuthenticationMethod)
+					.collect(Collectors.collectingAndThen(Collectors.toCollection(HashSet::new), Collections::unmodifiableSet));
```

Comment:
This is really nice! Unfortunately, we will want to [avoid the use of streams](https://github.com/spring-projects/spring-security/issues/9642).

Perhaps a new private method called from inside `build()` similar to `validateScopes()` and `validateRedirectUris()`:

```java
		private void upgradeClientAuthenticationMethods() {
			if (this.clientAuthenticationMethods.contains(ClientAuthenticationMethod.BASIC)) {
				this.clientAuthenticationMethods.remove(ClientAuthenticationMethod.BASIC);
				this.clientAuthenticationMethods.add(ClientAuthenticationMethod.CLIENT_SECRET_BASIC);
			}
			if (this.clientAuthenticationMethods.contains(ClientAuthenticationMethod.POST)) {
				this.clientAuthenticationMethods.remove(ClientAuthenticationMethod.POST);
				this.clientAuthenticationMethods.add(ClientAuthenticationMethod.CLIENT_SECRET_POST);
			}
		}
```

Reply:
I have removed the stream api